lmdeploy/vl/model/internvl.py

```diff
@@ -6,9 +6,71 @@
 from PIL.Image import Image
 from transformers import AutoConfig, AutoModel, CLIPImageProcessor
 
+from lmdeploy.utils import get_logger
 from lmdeploy.vl.model.base import VisonModel
 from lmdeploy.vl.model.utils import load_model_from_weight_files
 
+logger = get_logger('lmdeploy')
+
+
+def find_closest_aspect_ratio(aspect_ratio, target_ratios, width, height,
+                              image_size):
+    best_ratio_diff = float('inf')
+    best_ratio = (1, 1)
+    area = width * height
+    for ratio in target_ratios:
+        target_aspect_ratio = ratio[0] / ratio[1]
+        ratio_diff = abs(aspect_ratio - target_aspect_ratio)
+        if ratio_diff < best_ratio_diff:
+            best_ratio_diff = ratio_diff
+            best_ratio = ratio
+        elif ratio_diff == best_ratio_diff:
+            if area > 0.5 * image_size * image_size * ratio[0] * ratio[1]:
+                best_ratio = ratio
+    return best_ratio
+
+
+def dynamic_preprocess(image,
+                       min_num=1,
+                       max_num=6,
+                       image_size=448,
+                       use_thumbnail=False):
```

Comment:
lack docstring or type hint. Besides, may add a link to the original implementation to explain why adding the two functions.